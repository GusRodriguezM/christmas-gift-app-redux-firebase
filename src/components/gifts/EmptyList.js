import React from 'react';
import { Empty } from '../styles/gifts/list/List.styled';

export const EmptyList = () => {
    return (
        <Empty>There is no gifts in your list. Please start adding some!</Empty>
    )
}