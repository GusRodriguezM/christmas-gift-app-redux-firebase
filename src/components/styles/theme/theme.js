const theme = {
    colors: {
        black: '#000000',
        loginButton: '#4285f4',
        disabledButton: '#999595',
        red: '#C30F16',
        white: '#FFFFFF',
        list: '#c8e4e4',
        navbar: '#e8ceac',
        boxShadow: 'rgba(9, 9, 9, 0.23)',
    }
}

export default theme;