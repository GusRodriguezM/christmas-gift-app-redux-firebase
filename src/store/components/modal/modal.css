.modal {
    position: fixed;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;
    background-color: rgba(0, 0, 0, 0.5)    ;
    display: flex;
    align-items: center;
    justify-content: center;
}

.modal__content {
    width:  40%;
    height: 90%;
    background-color: #c8e4e4;
}

.modal__content--header, .modal__content--footer {
    padding: 10px;
    height: 10%;
}

.modal__content--title {
    margin: 0;
}

.modal__content--body {
    padding: 10px;
    border-top: 1px solid #c8e4e4;
    border-bottom: 1px solid #c8e4e4;
    height: 80%;
}