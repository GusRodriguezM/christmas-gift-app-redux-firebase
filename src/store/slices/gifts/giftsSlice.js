import { createSlice } from '@reduxjs/toolkit';

export const giftsSlice = createSlice({
  name: 'gifts',
  initialState: {
    gifts: [
      {
        name: 'Stuffed Bear',
        quantity: 1,
        image: 'https://images.unsplash.com/photo-1556012018-50c5c0da73bf?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=870&q=80',
        person: 'Janet',
        price: 10,
      },
      {
        name: 'Xbox Series S',
        quantity: 1,
        image: 'https://images.unsplash.com/photo-1605901309584-818e25960a8f?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1019&q=80',
        person: 'Jorge',
        price: 20,
      },
      {
        name: 'Guitar',
        quantity: 1,
        image: 'https://images.unsplash.com/photo-1516924962500-2b4b3b99ea02?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=870&q=80',
        person: 'Victor',
        price: 30,
      }
    ]
  },
  reducers: {
    addGift: (state) => { },
  },
});

export const { addGift } = giftsSlice.actions;