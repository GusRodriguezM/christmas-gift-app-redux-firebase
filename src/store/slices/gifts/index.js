export * from './giftsSlice';
export * from './thunks';