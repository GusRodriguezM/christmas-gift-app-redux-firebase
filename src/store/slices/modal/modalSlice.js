import { createSlice } from '@reduxjs/toolkit';

export const modalSlice = createSlice({
  name: 'modal',
  initialState: {
    open: false,
    option: ''
  },
  reducers: {
    openModal: (state) => {
        state.open = true;
    },
    closeModal: (state) => {
        state.open = false;
    },
    setOption: (state, action) => {
      state.option = action.payload;
    }
  },
});

export const { openModal, closeModal, setOption } = modalSlice.actions;