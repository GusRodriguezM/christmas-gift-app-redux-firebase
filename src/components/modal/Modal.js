import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { deleteActiveGift, deleteImageContent, resetSavingGift } from '../../store/slices/gifts';
import { closeModal } from '../../store/slices/modal/modalSlice';
import { ContentBody, HeaderFooter, ModalContainer, ModalContent } from '../styles/modal/ModalContainer.styled';
import { GiftButton } from '../styles/shared/Button.styled';

export const Modal = ({title, children}) => {

    const { open } = useSelector( state => state.modal );
    const dispatch = useDispatch();

    const handleCloseModal = () => {
        dispatch( closeModal() );
        dispatch( deleteActiveGift() );
        dispatch( deleteImageContent() );
        dispatch( resetSavingGift() );
    }

    if(!open){
        return null;
    }

    return (
        <ModalContainer onClick={handleCloseModal}>
            <ModalContent onClick={e => e.stopPropagation()}>
                <HeaderFooter>
                    <h3>{title}</h3>
                </HeaderFooter>
                
                <ContentBody>
                    {children}
                </ContentBody>

                <HeaderFooter>
                    <GiftButton
                        onClick={handleCloseModal}
                    >
                        <span>Close</span>
                        <i className="fa-solid fa-circle-xmark"></i>
                    </GiftButton>
                </HeaderFooter>
            </ModalContent>
        </ModalContainer>
    )
}