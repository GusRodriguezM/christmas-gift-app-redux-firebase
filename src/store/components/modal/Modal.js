import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { deleteActiveGift } from '../../slices/gifts';
import { closeModal } from '../../slices/modal/modalSlice';

import './modal.css'

export const Modal = ({title, children}) => {

    const { open } = useSelector( state => state.modal );
    const dispatch = useDispatch();

    const handleCloseModal = () => {
        dispatch( closeModal() );
        dispatch( deleteActiveGift() );
    }

    if(!open){
        return null;
    }

    return (
        <div className='modal' onClick={handleCloseModal}>
            <div className='modal__content' onClick={e => e.stopPropagation()}>
                <div className='modal__content--header'>
                    <h3 className='modal__content--title'>
                        {title}
                    </h3>
                </div>
                
                <div className='modal__content--body'>
                    {children}
                </div>

                <div className='modal__content--footer'>
                    <button
                        onClick={handleCloseModal}
                        className='button'
                    >
                        <span>Close</span>
                    </button>
                </div>
            </div>
        </div>
    )
}