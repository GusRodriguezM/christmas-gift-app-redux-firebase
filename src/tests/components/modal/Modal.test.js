import { configureStore } from "@reduxjs/toolkit";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { Modal } from "../../../components/modal/Modal";
import theme from "../../../components/styles/theme/theme";
import { deleteActiveGift, deleteImageContent, resetSavingGift } from "../../../store/slices/gifts/giftsSlice";
import { closeModal, modalSlice } from "../../../store/slices/modal/modalSlice";
import { initialState, openModalToAddGiftState } from "../../fixtures/modalFixtures";

//Mocking the functions used by the modal
const mockCloseModal = jest.fn();
const mockDeleteActiveGift = jest.fn();
const mockDeleteImageContent = jest.fn();
const mockResetSavingGift = jest.fn();

//Making a mock of the closeModal only
jest.mock('../../../store/slices/modal/modalSlice', () => ({
    ...jest.requireActual('../../../store/slices/modal/modalSlice'),
    closeModal: () => mockCloseModal
}));

jest.mock('../../../store/slices/gifts/giftsSlice', () => ({
    deleteActiveGift: () => mockDeleteActiveGift,
    deleteImageContent: () => mockDeleteImageContent,
    resetSavingGift: () => mockResetSavingGift
}));

describe('Tests in the <Modal /> component', () => {
    
    //Variables to send to the modal
    const title = 'Testing';
    const content = 'Content in the modal';

    //Mocking the store with the modalSlice
    const getMockStore = ( initState ) => {
        return configureStore({
            reducer: {
                modal: modalSlice.reducer
            },
            preloadedState: {
                modal: initState
            }
        });
    }
    
    test('Should not render the component if open is equal to false', () => {

        //Getting the store
        const store = getMockStore( {...initialState} );

        //Rendering the component
        //This should not render the modal
        const { container } = render(
            <Provider store={ store } >
                <ThemeProvider theme={ theme } >
                    <Modal title={ title } >
                        <h1>{content}</h1>
                    </Modal>
                </ThemeProvider>
            </Provider>
        );

        //Expecting that the container matches with the snapshot
        expect( container ).toMatchSnapshot();
    
    });

    test('Should render the component if open is equal to true', () => {

        //Getting the store
        const store = getMockStore( {...openModalToAddGiftState} );

        //Rendering the component
        const { container } = render(
            <Provider store={ store } >
                <ThemeProvider theme={ theme } >
                    <Modal title={ title } >
                        <h1>{content}</h1>
                    </Modal>
                </ThemeProvider>
            </Provider>
        );

        //Expecting that the container matches with the snapshot
        expect( container ).toMatchSnapshot();
        
        //If the above is ok then in the modal should appear the h tags for the title and content
        expect( screen.getByRole('heading', { level: 3, name: `${title}` }) ).toBeTruthy();
        expect( screen.getByRole('heading', { level: 1, name: `${content}` }) ).toBeTruthy();

    });

    test('Close button should call closeModal, deleteActiveGift, deleteImageContent, resetSavingGift actions', () => {

        //Getting the store
        const store = getMockStore( {...openModalToAddGiftState} );

        //Rendering the component
        render(
            <Provider store={ store } >
                <ThemeProvider theme={ theme } >
                    <Modal title={ title } >
                        <h1>{content}</h1>
                    </Modal>
                </ThemeProvider>
            </Provider>
        );

        //Getting the reference to the button
        const closeBtn = screen.getByRole('button', { name: 'Close' });
        
        //Firing the event
        fireEvent.click( closeBtn );

        //Expecting that the actions have been called after clicking the button
        expect( mockCloseModal ).toHaveBeenCalled();
        expect( mockDeleteActiveGift ).toHaveBeenCalled();
        expect( mockDeleteImageContent ).toHaveBeenCalled();
        expect( mockResetSavingGift ).toHaveBeenCalled();

    });

});