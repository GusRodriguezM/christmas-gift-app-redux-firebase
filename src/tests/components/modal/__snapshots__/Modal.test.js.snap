// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in the <Modal /> component Should not render the component if open is equal to false 1`] = `<div />`;

exports[`Tests in the <Modal /> component Should render the component if open is equal to true 1`] = `
<div>
  <div
    class="sc-bczRLJ eFUSeP"
  >
    <div
      class="sc-gsnTZi geZUCt"
    >
      <div
        class="sc-dkzDqf hkPnie"
      >
        <h3>
          Testing
        </h3>
      </div>
      <div
        class="sc-hKMtZM bcpxhA"
      >
        <h1>
          Content in the modal
        </h1>
      </div>
      <div
        class="sc-dkzDqf hkPnie"
      >
        <button
          class="sc-eCYdqJ sc-jSMfEi sEdKe hmovcX"
        >
          <span>
            Close
          </span>
          <i
            class="fa-solid fa-circle-xmark"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
